java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -86,7 +88,38 @@
     private final KinesisProducerConfiguration config;
     private final Map<String, String> env;
     private final AtomicLong messageNumber = new AtomicLong(1);
-    private final Map<Long, SettableFuture<?>> futures = new ConcurrentHashMap<>();
+    private final Map<Long, SettableFutureTracker> futures = new ConcurrentHashMap<>();
+    private final PriorityQueue<SettableFutureTracker> oldestFutureTrackerHeap = new PriorityQueue(new SettableFutureTrackerComparator());
```

Comment:
Given that this is doubling the size of in-memory objects, can we make this configurable?

Reply:
Will discuss offline but what i feel is that we should actually not have this additional heap and keep the getOldestRecordTimeInSeconds implementation to be O(N) which is fine since that is something customer application will invoke and they can control and frequency. Optimizing that method to be O 1 seems un-needed efficiency by sacrificing additional memory usage.